File: CSharp - Monogame SDK 3.7/Astro.cs
using System.Collections.Generic;
using System.IO;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Audio;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using Microsoft.Xna.Framework.Media;

namespace Asteroids {

    public sealed class Astro : Game {

        // For this simple game, implementing Astro as a singleton is an alternative to
        // having its public member variables be static -- both approaches provide an easy
        // way for the entities to reference "global" info. Not sure which is "better,"
        // but it would also be possible to simply create a normal instance of Astro in Program
        // as a global variable and call it a day.
        private static Astro instance = null;
        public static Astro Instance { get => instance ?? (instance = new Astro()); }

        public readonly int WIDTH = 1280;
        public readonly int HEIGHT = 720;
        public readonly Dictionary<string, SoundEffect> sounds = new Dictionary<string, SoundEffect>();
        public readonly Dictionary<string, Animation> animations = new Dictionary<string, Animation>();
        public readonly List<Entity> entities = new List<Entity>();

        public enum GameState { READY, PLAYING, GAMEOVER };
        public GameState gameState = GameState.READY;
        public List<Entity> currentEntities;
        public int score;
        public Ship ship;     // the ship needs to be accessed by the saucer
        public Saucer saucer; // and the saucer needs to be accessed by the ship

        KeyboardState kbState, kbPrevState;
        ButtonState btnState, btnPrevState;
        SpriteBatch spriteBatch;
        SpriteFont font;
        Texture2D spritesheet, background;
        string playerName;

        private Astro() {
            new GraphicsDeviceManager(this) {
                GraphicsProfile = GraphicsProfile.HiDef, // enable support for large textures
                PreferredBackBufferWidth = WIDTH, // window dimensions
                PreferredBackBufferHeight = HEIGHT
            };
            Content.RootDirectory = "Content";
            Window.TextInput += HandleTextInput; // add event handler
        }

        protected override void LoadContent() {
            spriteBatch = new SpriteBatch(GraphicsDevice);
            font = Content.Load<SpriteFont>("OCR");

            System.Uri uri = new System.Uri("audio/background.mp3", System.UriKind.Relative);
            MediaPlayer.IsRepeating = true;
            MediaPlayer.Volume = 0.02f;
            MediaPlayer.Play(Song.FromUri("background", uri));

            sounds["bang_lg"] = LoadSound("bang_lg");
            sounds["bang_md"] = LoadSound("bang_md");
            sounds["bang_sm"] = LoadSound("bang_sm");
            sounds["laser"] = LoadSound("laser");
            sounds["thrust"] = LoadSound("thrust");
            sounds["saucer"] = LoadSound("saucer");

            background = LoadTexture("outerspace.jpg");
            spritesheet = LoadTexture("spritesheet.png");
            animations["asteroid_lg"] = new Animation(spritesheet, 0, 384, 80, 80, 16, 0.15);
            animations["asteroid_md"] = new Animation(spritesheet, 0, 464, 60, 60, 16, 0.15);
            animations["asteroid_sm"] = new Animation(spritesheet, 0, 524, 40, 40, 16, 0.15);
            animations["explosion_asteroid"] = new Animation(spritesheet, 0, 0, 192, 192, 32, 0.4);
            animations["explosion_ship"] = new Animation(spritesheet, 0, 192, 192, 192, 32, 0.4);
            animations["laser"] = new Animation(spritesheet, 1540, 384, 32, 48, 8, 0.4);
            animations["laser_saucer"] = new Animation(spritesheet, 1540, 444, 32, 48, 8, 0.4);
            animations["saucer"] = new Animation(spritesheet, 2200, 384, 45, 73, 6, 0.15);
            animations["ship"] = new Animation(spritesheet, 1820, 384, 42, 140, 1, 0, -34);
            animations["ship_thrusting"] = new Animation(spritesheet, 1862, 384, 42, 140, 6, 0.25, -34);

            HighScores.Instance.ReadWeb();
        }

        SoundEffect LoadSound(string filename) {
            using (FileStream fileStream = new FileStream($"audio/{filename}.wav", FileMode.Open)) {
                return SoundEffect.FromStream(fileStream);
            }
        }

        Texture2D LoadTexture(string filename) {
            using (FileStream fileStream = new FileStream($"images/{filename}", FileMode.Open)) {
                return Texture2D.FromStream(GraphicsDevice, fileStream);
            }
        }

        void InitGame() {
            score = 0;
            playerName = "";
            foreach (Entity e in entities) e.active = false;
            entities.Clear();
            ship = new Ship(WIDTH * 0.5f, HEIGHT * 0.5);
            saucer = new Saucer(); // there is only one saucer, but usually it is hidden
            entities.Add(ship);
            entities.Add(saucer);
            for (int i = 0; i < 3; i++)
                Asteroid.Spawn(animations["asteroid_lg"], 0, Program.random.Next(HEIGHT));
        }

        void HandleTextInput(object sender, TextInputEventArgs e) {
            if (gameState == GameState.READY) {
                InitGame();
                gameState = GameState.PLAYING;
                return;
            }
            if (gameState == GameState.PLAYING) return; // no text input during game play

            if (e.Character == '\r' && playerName.Length > 0) {
                HighScores.Instance.WriteWeb(playerName, score.ToString());
                HighScores.Instance.ReadWeb();
                gameState = GameState.READY;
            }
            else if (e.Character == '\b' && playerName.Length != 0) {
                playerName = playerName.Substring(0, playerName.Length - 1);
            }
            else if ((int)e.Character > 32 && playerName.Length < 10) {
                playerName += e.Character;
            }
        }

        void HandleInput() {
            kbState = Keyboard.GetState();
            if (kbState.IsKeyDown(Keys.Escape)) Exit();
            if (kbState.IsKeyDown(Keys.Space) && kbPrevState.IsKeyUp(Keys.Space)) {
                Laser.Spawn(ship.x, ship.y, ship.heading);
                sounds["laser"].Play();
            }
            if (kbState.IsKeyDown(Keys.Z)) // lots 'o lasers!!!!
                Laser.Spawn(ship.x, ship.y, ship.heading);
            if (kbState.IsKeyDown(Keys.A) || kbState.IsKeyDown(Keys.Left))
                ship.heading -= 0.075;
            if (kbState.IsKeyDown(Keys.D) || kbState.IsKeyDown(Keys.Right))
                ship.heading += 0.075;
            ship.thrusting = (kbState.IsKeyDown(Keys.W) || kbState.IsKeyDown(Keys.Up));
            kbPrevState = kbState;

            btnState = Mouse.GetState().LeftButton;
            if (btnState == ButtonState.Released && btnPrevState == ButtonState.Pressed) {
                Laser.Spawn(ship.x, ship.y, ship.heading);
                sounds["laser"].Play();
            }
            btnPrevState = btnState;
        }

        protected override void Update(GameTime gameTime) {
            if (gameState == GameState.PLAYING) {
                HandleInput();

                if (Program.random.Next() % 750 == 0)
                    Asteroid.Spawn(animations["asteroid_lg"], 0, Program.random.Next(HEIGHT));
                if (Program.random.Next() % 900 == 0 && saucer.hidden) {
                    saucer.Show();
                }

                // loop thru a copy of the entities list so we can modify the original
                currentEntities = new List<Entity>(entities);
                foreach (Entity e in currentEntities) {
                    e.Update();
                    e.CheckCollisions();
                    if (!e.active) entities.Remove(e);
                }
            }
            base.Update(gameTime);
        }

        protected override void Draw(GameTime gameTime) {
            spriteBatch.Begin(SpriteSortMode.Deferred, BlendState.NonPremultiplied);
            spriteBatch.Draw(background, Vector2.Zero, Color.White);
            if (gameState == GameState.READY) {
                HighScores.Instance.DrawWeb(spriteBatch, font);
                spriteBatch.DrawString(font, "- Press enter to begin -", new Vector2(420, 550), Color.White);
            }
            if (gameState == GameState.PLAYING || gameState == GameState.GAMEOVER) {
                foreach (Entity s in entities) s.Draw(spriteBatch);
                ship.DrawExtraShips(spriteBatch);
                spriteBatch.DrawString(font, $"Score: {score}" , new Vector2(20, 20), Color.White);
            }
            if (gameState == GameState.GAMEOVER) {
                spriteBatch.DrawString(font, "You set a new high score!", new Vector2(420, 150), Color.Wheat);
                spriteBatch.DrawString(font, "Name:", new Vector2(420, 220), Color.White);
                spriteBatch.DrawString(font, playerName, new Vector2(520, 220), Color.White);
                if (gameTime.TotalGameTime.Milliseconds % 500 > 250)
                    spriteBatch.DrawString(font, "|", new Vector2(510 + 20 * playerName.Length, 220), Color.Cyan);
            }
            spriteBatch.End();
            base.Draw(gameTime);
        }
    }
}

File: TypeScript - p5/src/Astro.ts
import P5 from "p5";
//HACK: for p5@1.3.1, need to modify p5/lib/addons/p5.sound.js and add 'import p5 from "p5";' to the top
import "p5/lib/addons/p5.sound";
import Animation from "./Animation";
import Asteroid from "./Asteroid";
import Entity from "./Entity";
import HighScores from "./HighScores";
import Laser from "./Laser";
import Saucer from "./Saucer";
import Ship from "./Ship";

export enum GameState {
  READY,
  PLAYING,
  GAMEOVER,
}

export default class Astro {
  static readonly WIDTH = 1200;
  static readonly HEIGHT = 720;
  static readonly MAXSPEED = 15;
  static readonly SPACEBAR = 32;
  static readonly LOWER_Z = 90;
  static readonly NEW_ASTEROID_FREQUENCY = 800; // the odds of a new asteroid appearing are 1 in 800
  static readonly NEW_SAUCER_FREQUENCY = 900;
  static playerName = "";
  static score = 0;
  static prevKeyIsDown: Record<any, boolean> = { SPACEBAR: false }; // initialize previous state of the keys (we only care about spacebar)
  static sounds: Record<string, P5.SoundFile> = {};
  static animations: Record<string, Animation> = {};
  static entities: Entity[] = []; // all the active game objects= ship; saucer; asteroids; lasers; explosions
  static currentEntities: Entity[] = []; // a copy used to loop; since the real entities array is modified during the loop
  static gameState = GameState.READY;
  static spritesheet: P5.Image;
  static background: P5.Image;
  static font: P5.Font;
  static ship: Ship; // the ship and saucer objects need references to each other
  static saucer: Saucer | null = null;
}

export function loadContent(p5: P5) {
  // assets sourced from original FamTrinli project files
  // additional assets from https://opengameart.org
  Astro.spritesheet = p5.loadImage("images/spritesheet.png");
  Astro.background = p5.loadImage("images/outerspace.jpg");
  Astro.font = p5.loadFont("fonts/arcade_i.ttf"); // https://www.dafont.com/arcade-ya.font

  //HACK: for p5@1.3.1, need to cast p5 as a SoundFile to access loadSound()
  const loadSound = (path: string) => (p5 as unknown as P5.SoundFile).loadSound(path);
  Astro.sounds["bang_lg"] = loadSound("audio/bang_lg.wav");
  Astro.sounds["bang_md"] = loadSound("audio/bang_md.wav");
  Astro.sounds["bang_sm"] = loadSound("audio/bang_sm.wav");
  Astro.sounds["laser"] = loadSound("audio/laser.wav");
  Astro.sounds["thrust"] = loadSound("audio/thrust.wav");
  Astro.sounds["saucer"] = loadSound("audio/saucer.wav");
  // https://patrickdearteaga.com/arcade-music/
  Astro.sounds["background"] = loadSound("audio/solveThePuzzle.ogg");
}

export function setupSounds() {
  Astro.sounds["thrust"].playMode("untilDone");
  Astro.sounds["saucer"].setVolume(0.1);
  Astro.sounds["background"].setVolume(0.05);
  Astro.sounds["background"].loop();
}

export function createAnimations() {
  // textures can be re-used for multiple animations; define which parts of each texture to use for each sprite
  Astro.animations["asteroid_lg"] = new Animation(Astro.spritesheet, 0, 384, 80, 80, 16, 0.15);
  Astro.animations["asteroid_md"] = new Animation(Astro.spritesheet, 0, 464, 60, 60, 16, 0.15);
  Astro.animations["asteroid_sm"] = new Animation(Astro.spritesheet, 0, 524, 40, 40, 16, 0.15);
  Astro.animations["explosion_asteroid"] = new Animation(Astro.spritesheet, 0, 0, 192, 192, 32, 0.4);
  Astro.animations["explosion_ship"] = new Animation(Astro.spritesheet, 0, 192, 192, 192, 32, 0.4);
  Astro.animations["laser"] = new Animation(Astro.spritesheet, 1540, 384, 32, 48, 8, 0.4);
  Astro.animations["laser_saucer"] = new Animation(Astro.spritesheet, 1540, 444, 32, 48, 8, 0.4);
  Astro.animations["saucer"] = new Animation(Astro.spritesheet, 2200, 384, 45, 73, 6, 0.15);
  Astro.animations["ship"] = new Animation(Astro.spritesheet, 1820, 384, 42, 140, 1, 0, 34);
  Astro.animations["ship_thrusting"] = new Animation(Astro.spritesheet, 1862, 384, 42, 140, 6, 0.25, 34);
}

function initGame() {
  Astro.score = 0;
  Astro.playerName = "";
  Astro.entities.splice(0); // clear the entities array
  Astro.ship = new Ship(Astro.animations["ship"], Astro.WIDTH * 0.5, Astro.HEIGHT * 0.5);
  Astro.entities.push(Astro.ship);
  // create some initial asteroids
  for (let i = 0; i < 3; i++) Asteroid.spawn(Astro.animations["asteroid_lg"], 0, Math.random() * Astro.HEIGHT);
}

function readyEvents(p5: P5) {
  if (p5.keyIsPressed) {
    initGame();
    Astro.gameState = GameState.PLAYING;
  }
}

function playingEvents(p5: P5) {
  if (p5.keyIsDown(Astro.SPACEBAR) && !Astro.prevKeyIsDown[Astro.SPACEBAR]) {
    // only fire when the spacebar changes state
    Laser.spawn(Astro.animations["laser"], Astro.ship.x, Astro.ship.y, Astro.ship.heading);
    Astro.sounds["laser"].play();
  }
  Astro.prevKeyIsDown[Astro.SPACEBAR] = p5.keyIsDown(Astro.SPACEBAR);

  Astro.ship.thrusting = p5.keyIsDown(p5.UP_ARROW);
  if (p5.keyIsDown(p5.LEFT_ARROW)) Astro.ship.heading -= 0.08; // rotate clockwise
  if (p5.keyIsDown(p5.RIGHT_ARROW)) Astro.ship.heading += 0.08;
  if (p5.keyIsDown(Astro.LOWER_Z))
    Laser.spawn(Astro.animations["laser"], Astro.ship.x, Astro.ship.y, Astro.ship.heading); // lots 'o lasers!!!!
}

function gameoverEvents() {
  // we just use the keyRelased() function to handle gameover events
}

export function handleEvents(p5: P5) {
  switch (Astro.gameState) {
    case GameState.READY:
      readyEvents(p5);
      break;
    case GameState.PLAYING:
      playingEvents(p5);
      break;
    case GameState.GAMEOVER:
      gameoverEvents();
      break;
  }
  if (p5.keyIsDown(p5.ESCAPE)) {
    Astro.sounds["background"].stop();
    Astro.sounds["saucer"].stop();
    p5.noLoop();
  }
}

export function updateEntities(p5: P5) {
  if (Astro.gameState !== GameState.PLAYING) return; // only update entities during game play

  if (Math.floor(Math.random() * Astro.NEW_ASTEROID_FREQUENCY) === 0)
    // throw in another asteroid occasionally
    Asteroid.spawn(Astro.animations["asteroid_lg"], 0, Math.random() * Astro.HEIGHT);

  if (Math.floor(Math.random() * Astro.NEW_SAUCER_FREQUENCY) === 0 && Astro.saucer === null) {
    // throw in a flying saucer occasionally
    Astro.saucer = new Saucer(Astro.animations["saucer"]); // we need a global reference
    Astro.entities.push(Astro.saucer);
  }

  // loop thru a copy of the entities list so we can modify the original
  Astro.currentEntities = [...Astro.entities]; // make a copy of the entities array
  for (let e of Astro.currentEntities) {
    e.update(); // update the heading, velocity, and/or position of the entity
    e.checkCollisions();
    if (!e.active) Astro.entities.splice(Astro.entities.indexOf(e), 1); // remove inactive entity
  }
}

export function drawEntities(p5: P5) {
  p5.image(Astro.background, Astro.WIDTH * 0.5, Astro.HEIGHT * 0.5); // draw background image centered on the canvas

  if (Astro.gameState === GameState.READY) {
    HighScores.draw(p5);
    p5.text("- Press enter to play -", 340, 570);
  }

  if (Astro.gameState === GameState.PLAYING || Astro.gameState === GameState.GAMEOVER) {
    for (let e of Astro.entities) e.draw(p5);
    Astro.ship.drawExtraShips(p5);
    p5.text("Score:" + Astro.score, 20, 40);
  }

  if (Astro.gameState === GameState.GAMEOVER) {
    p5.text("You set a new high score!", 420, 150);
    p5.text(`Name: ${Astro.playerName}`, 420, 220);
    if (new Date().getMilliseconds() % 500 > 250)
      // "blink" the cursor every quarter second or so
      p5.text("*", 528 + 18 * Astro.playerName.length, 220);
  }
}
